bitcoin/src/taproot/merkle_branch/mod.rs

```diff
@@ -1,9 +1,62 @@
 //! Contains `TaprootMerkleBranchBuf` and its associated types.
 
 mod buf;
+mod borrowed;
 
 pub use buf::TaprootMerkleBranchBuf;
+pub use borrowed::TaprootMerkleBranch;
+
+use core::fmt;
 use super::{
     InvalidMerkleBranchSizeError, InvalidMerkleTreeDepthError, TapNodeHash, TaprootError,
     TAPROOT_CONTROL_MAX_NODE_COUNT, TAPROOT_CONTROL_NODE_SIZE,
 };
+
+/// Returned when decoding of merkle branch fails.
+#[derive(Debug)]
+pub struct DecodeError {
+    // To save space we don't carry around node count but byte count
+    // This way we can represent both error kinds in the same `usize` rather than blowing it up to
+    // twice the size. Combined with niche optimization on references this should get us `Result`
+    // with the same sized as the returned type.
+    num_bytes: usize,
```

Comment:
In d480bb4edf7061abde65058c0278091ce6900bb1:

Hah, clever. I think the comment needs to be updated though to reduce the amount of bragging about the optimization and increase the amount of explaining what is going on. e.g.

```
/// Represents the invalid number of bytes. It may be invalid in two ways: it might not
/// be a multiple of 32, in which case it is guaranteed to be wrong for that reason;
/// only if it is a multiple of 32 do we check that it does not exceed 32 * 128, in which
/// case it is wrong for that reason.
///
/// This error type is used in `Result<&T, DecodeError>`, so by keeping its size down to
/// a single `usize` and considering the niche optimization on `&T`, the result will
/// have the same size as just `&T`.
```

Reply:
Sure, I'll just need to point out that `T` must be unsized for that claim to be true. (Also it needs sufficiently new Rust version to have effect, IIRC it was 70-something but I don't think it needs being mentioned.)